pkg/controller/podautoscaler/horizontal.go

```diff
@@ -422,6 +422,15 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 		case desiredReplicas > scaleUpLimit:
 			setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "ScaleUpLimit", "the desired replica count is increasing faster than the maximum scale rate")
 			desiredReplicas = scaleUpLimit
+
+			// Ensure that even if the scaleUpLimit is greater
+			// than the maximum number of replicas, we only
+			// set the max number of replicas as desired.
+			if scaleUpLimit > hpa.Spec.MaxReplicas {
```

Comment:
This is basically an exact copy-paste of one of the cases below. Adding ifs checking whether we would trigger condition of a different case if we haven't already triggered the one we're in makes me think switch-case is not the right solution here and we should move to a sequence of ifs instead (perhaps contained in a separate function that can be unittested in isolation).

Especially since we already had a very similar bug caused by interaction of conditions in this switch-case: https://github.com/kubernetes/kubernetes/pull/48997.

I'm ok with keeping this bugfix as is, but we should revisit this part of hpa, clean it up and make sure it's reasonably unittested to avoid another such issue in future. @mwielgus @DirectXMan12

Reply:
I've created an issue #53728 to do this.